### ISSUE TITLE
Create DefaultInterceptor which will create & dispose service provider scopes, if one is not set.

### ISSUE THREAD

**Issue by ajohnstone-ks** [MarimerLLC/csla#1996]
**Is your feature request related to a problem? Please describe.**
Csla should include a default interceptor which will check if a ServiceProviderScope has been set in Initialize.  If it has, it will do nothing.  If not, it will create one, and dispose of it during Complete.

This will ensure that server side disposable transients configured in a DI container are properly disposed before the data portal call returns.

**Describe the solution you'd like**
An interceptor which will create a service provider scope from the DefaultServiceProvider is none is set, and dispose it when Completed.  The interceptor should be used unless a different one is configured, except for a remote application host scenario.

See this discussion: https://github.com/MarimerLLC/csla/discussions/1983

**Comment by ajohnstone-ks** [comment]
using Microsoft.Extensions.DependencyInjection;

namespace Csla.Server
{
    /// <summary>The default interceptor will create a new <see cref="ApplicationContext.ServiceProviderScope"/>
    /// on <see cref="IInterceptDataPortal.Initialize"/> and dispose of it on <see cref="IInterceptDataPortal.Complete"/>.</summary>
    public class DefaultInterceptor : IInterceptDataPortal
    {
      private IServiceScope _scope;
      
      public virtual void Initialize(InterceptArgs e) { }

      public virtual void Complete(InterceptArgs e) { }

      void IInterceptDataPortal.Initialize(InterceptArgs e) {
        if (ApplicationContext.ServiceProviderScope == null && ApplicationContext.DefaultServiceProvider != null)
        {
          var scopeFactory = ApplicationContext.DefaultServiceProvider.GetRequiredService<IServiceScopeFactory>();
          _scope = scopeFactory.CreateScope();
          ApplicationContext.ServiceProviderScope = _scope.ServiceProvider;
        }

        Initialize(e);
      }
      
      void IInterceptDataPortal.Complete(InterceptArgs e)
      {
        Complete(e);

        if (_scope != null)
        {
          ApplicationContext.ServiceProviderScope = null;
          _scope.Dispose();
        }
      }
    }
}

**Comment by rockfordlhotka** [comment]
As per discussion in #1999, this change should include creating a `InterceptorBase` type that exposes protected properties for the default and current DI provider. That will ease implementation of custom interceptors that need those values.

**Comment by dazinator** [comment]
Is it worth making the change to activate this by DI, or would that come later down the road? The reason I ask is because, if DI is used to activate this type, then the user can have the `DefaultServiceProvider` injected via the constructor, and a base type would not be necessary, as `ApplicationContext.ServiceProviderScope` can still then be checked and set if necessary?

**Comment by rockfordlhotka** [comment]
Ahh, yes, I see what you are talking about. So we probably _don't_ need the base type, which is excellent!

All of this is for CSLA 6, and CSLA 6 will require DI and will be creating types using the DI activator, so such injection should work fine.

**Comment by ajohnstone-ks** [comment]
@dazinator @rockfordlhotka 

How is this envisioned to work exactly?  How (or will) Csla enforce that the registration for `IServiceProvider` always return the root SP?  Will it use DefaultServiceProvider always to build the interceptor, or will it use CurrentServiceProvider?  If it uses Default, I think it'd be hard to make a registration that didn't return the root, but if it uses Current, the interceptor instances might get the currently scoped SP depending on how IServiceProvider was registered.  Is one of the desired changes to allow Csla users to control the lifetime of interceptors by letting them register them as Singletons, transient, or some variation (transient unless my parent container has an instance, then get that one)?

**Comment by dazinator** [comment]
> How is this envisioned to work exactly? How (or will) Csla enforce that the registration for IServiceProvider always return the root SP? 

The proposal is currently to make CSLA use the root / default IServiceProvider, to activate this type and supply DI parameters into the constructor. This guarantees that the root IServiceProvider will be injected in the constructor if you have it as a contructor arg.

> If it uses Default, I think it'd be hard to make a registration that didn't return the root,

Not following quite what you mean by "make a registration" as there is no DI registration for the interceptor in this context (will come back to this point). If your implementation needs access to a dependency that is not from the root scope (that CSLA is activating in), you have some options:

1. Your interceptor, can access `ApplicationContext.ServiceProviderScope` to check for a current scoped provider and use that to obtain a dependency from that scope on the fly, if it needs a scoped dependency. Obviously some part of your application would need to be setting that before hand - appropriately for that to be useful.
2. Your interceptor, can have a factory injected that provides it access to the scoped IServiceProvider or service that it needs.

###Side Note: Chaining interceptors is also a pattern that becomes possible in this model, for example, interceptor A runs first, then interceptor B runs - here is an adhoc example - not sure whether this kind of thing actually becomes useful:

```
services.AddTransient<FirstInterceptor>():
Csla.SetCustomActivator(typeof(SecondInterceptor));

public class FirstInterceptor : IInterceptor
{
 private Func<IServiceScope> _createSope;
 private IServiceScope _scope;
 private IServiceScope _originalScope;

 public FirstInterceptor(Func<IServiceScope> createSope)
 {
   _createSope= createSope;
 }
 
 public Before() // sorry don;t have actual method names after
 {
      _scope = _createSope();
      _originalScope = ApplicationContext.ServiceProviderScope;
      ApplicationContext.ServiceProviderScope = _scope;
 }

 public Completed() // sorry don;t have actual method names after
  {
   ApplicationContext.ServiceProviderScope = _originalScope;
   _scope.Dispose();
  }

}

public class SecondInterceptor: IInterceptor
{
  private IInterceptor _first;
  
  public SecondInterceptor(FirstInterceptor first) 
  {
  _first = first;  
  }
  
  public Before() // sorry don;t have actual method names after
  {
  // run inner interceptor first
   _first.Before(); 
   // now we've got a weird scope put here by the first interceptor, lets use that to get a service
   var scopedSP = ApplicationContext.ServiceProviderScope;
  // do whatever you want: scopedSP 
    var oddScopeService = scopedSP.GetService<Foo>();
    // create another scope etc?
  }
  
  public Completed() // sorry don;t have actual method names after
  {
  
  }
}

```

> Is one of the desired changes to allow Csla users to control the lifetime of interceptors by letting them register them as Singletons, transient, or some variation (transient unless my parent container has an instance, then get that one)?

I'm open to that, as I think the current model you have explained, where CSLA creates an instance of the interceptor per DP call - if true, can be creating more garbage than necessary. It would be nice to be able to register the interceptor itself with DI (and therefore its own lifetime) and then have CSLA resolve it - in which case it could get back a transient instance each time, or it could be getting back a singleton instance - for those that can write an interceptor in a thread safe way and register it as a singleton this would reduce some of that garbage - however I think that change is larger and can potentially come later (as a seperate issue) ?

Is there a reason why activating the interceptor in root scope, would be an issue for you, and could you give an example if so as that would be helpful!

**Comment by ajohnstone-ks** [comment]
> The proposal is currently to make CSLA use the root / default IServiceProvider, to activate this type and supply DI parameters into the constructor. This guarantees that the root IServiceProvider will be injected in the constructor if you have it as a contructor arg.

So the idea is to do something like this (I don't remember exactly where the type is stored)
`var interceptor = serviceProvider.GetService(typeof(ApplicationConext.IntercetorType));`

It could have also been
`var interceptor = serviceProvider.GetService(typeof(IInterceptDataPortal));` which could also just eliminate the specific Interceptor configuration in Csla... just configure your container correctly, but this requires you configure the container explicitly.  I just didn't know which way things were going, which is why I asked.

> Not following quite what you mean by "make a registration" as there is no DI registration for the interceptor in this context (will come back to this point).

There doesn't have to be a registration for the interceptor, but there can be.  You could do (using Unity, since that's what I'm familar with):
`container.RegisterType<MyCustomInterceptor>(new SingletonLifetimeManager())`

> Let's not forget that currently, the type activated by CSLA can't have any dependencies at all, so allowing dependencies to be injected from root scope is a step forward.

I'm not arguing that, I do think it brings some benefit.  But lifetime management also comes with enabling DI, which was my next point.

> I'm open to that, it's something that could either be done now - or as a later enhancement? I think the current model you have explained, where CSLA creates an instance of the interceptor per DP call, is possibly creating more garbage than necessary.

Actually I was pointing out that using a container to resolve the interceptor automatically opens that possibility immediately (see the registration above).  My question was is that considered one of the goals, or did it end up being a side effect?  You can rely on the fact that today an Interceptor is created just after every DataPortal.Create/Fetch/Delete/Update/Execute and is completed & forgotten after.  Now that can be changed just be configuring how the interceptor is resolved in your container, and this can be done completely outside of Csla.

Maybe someone has a good reason to have a Singleton interceptor, or maybe they need to be more explicit about how it gets built.   Its more flexible, but that also means it can be screwed up more easily, and container configuration issues can be really tough to track down.

> Is there a reason why activating the interceptor in root scope, would be an issue for you in the interim?

No actually that's what I would personally prefer; although maybe there's someone out there that would like it to be the currently scoped one 😆 I'm trying to get a better understanding of what the plan is because I want to make sure that we can work with the direction things are heading.  This is one of the most exciting features Csla has added IMO, but making sure I can do the scoping as needed so resources get released when we need them to is critical, and more so if Csla will required DI to be configured.

**Comment by dazinator** [comment]
> So the idea is to do something like this (I don't remember exactly where the type is stored)
var interceptor = serviceProvider.GetService(typeof(ApplicationConext.IntercetorType));

Sort of.. The use of GetService() in your above example would require that the type is also registered with the DI container with a lifetime such as transient, or Singleton etc . Im proposing we don't impose that just yet because:

1. That's an additional registration that users might not be expecting as it was never required before. A missing DI registration would cause a runtime error.
2. CSLA currently expects to create a new instance each time rather than share an instance. Changing CSLA to defer lifetime management to the container introduces the additional complication that a shared instance could be used (if registered as scoped or singleton) and that means additional guidance over how to safely create such a service (thread safety etc). I think this falls in the "nice to have" but not really essential category.

So basically - I think keep it simple for now by not requiring any DI registration, just have CSLA activate the instance (like it currently does) providing dependencies. Then only if there are concrete examples for why we need to go further and have CSLA delegate lifetime to the container then we can consider that change separately down the road and how it could be implemented to stop users making the wrong registration or registering singleton interceptors that have thread safety issues etc.

> No actually that's what I would personally prefer;

I think it makes sense too.

**Comment by rockfordlhotka** [comment]
I'm not adverse to the idea that CSLA 6 will require a lot of types currently provided through config to be registered via DI. This is a breaking change release, and I'd rather buy into DI now, than try to cobble together a nightmare combination of old-style provider models and new-style DI models. That will only lead to pain and confusion for years to come!

**Comment by dazinator** [comment]
@rockfordlhotka 

> I'm not adverse to the idea that CSLA 6 will require a lot of types currently provided through config to be registered via DI. This is a breaking change release, and I'd rather buy into DI now, than try to cobble together a nightmare combination of old-style provider models and new-style DI models. That will only lead to pain and confusion for years to come!
> 
> We won't get another chance for this sort of change until CSLA 7, which might not be for 2-3 years from now.

question then.. Interceptor instances are currently created by CSLA on demand for the DP operation in progress. Do you want the framework to delegate the lifetime of the interceptor to a DI registration so that the user is in control of whether an instance of there interceptor is always created, or is a singleton instance that is used concurrently? If you want this lifetime to be a choice for the user then that would force our hand - it would need to be registered with the DI container with a lifetime such as transient or Singleton.

If you want CSLA to remain in control of lifetime, by always creating a new instance as it currently does - then there are a couple of options:

1. Keep code similar to today by replacing Activator.CreateInstance with the alternative that uses the DI container to also inject dependencies when creating the instance. This is what I was proposing as it doesnt introduce any additional DI registration.
2. Have CSLA defer lifetime to the container by doing `var interceptorInstance = sp.GetService<IInterceptor>()` - users must register the interceptor with a lifetime that works e.g Transient lifetime would be in keeping with how CSLA works today.
I like option 2 in that csla would no longer require the concrete type to be configured as it could resolve by the contract.


The other benefit to forcing the user to register the interceptor as a service with DI is that they can register the interceptor with a factory function
```
services.AddTransient<MyInterceptor>(sp=> new MyInterceptor());
```

The current mechanism they can only provide the type and have dependencies injected which doesn't allow for a factory function to be used to create the instance - but that is a fairly minor limitation.

**Comment by rockfordlhotka** [comment]
Really that's a broader question across the whole set of providers that can be configured for use, and I would prefer that we come up with a consistent answer.

Right now, I am leaning toward moving all the provider types out of config and having them come from DI. I don't yet fully know the ramifications of that, especially around default providers, because people shouldn't have to configure the defaults, only do overrides where necessary.

Overall though, I'd prefer to switch all the provider type configuration to rely on DI.

**Comment by ajohnstone-ks** [comment]
> Sort of.. The use of GetService() in your above example would require that the type is also registered with the DI container with a lifetime such as transient, or Singleton etc . Im proposing we don't impose that just yet because:
> 
> 1. That's an additional registration that users might not be expecting as it was never required before. A missing DI registration would cause a runtime error.
> 2. CSLA currently expects to create a new instance each time rather than share an instance. Changing CSLA to defer lifetime management to the container (rather than always create an instance) introduces the additional complications that a shared instance could be used (if the user registered it as scoped or singleton instead of transient) and that means additional guidance over how to safely create such a service (to avoid thread safety issues etc). I think this falls in the "nice to have" but not really essential category. Its a viable option though if / when @rockfordlhotka would like to go down that road.

I think you're reading too much into what I am saying.  I wasn't advocating that Csla require an explicit registration with the container, I'm only stating that it immediately becomes a possibility, and I think the consequences of that do need to be considered.

I don't think Csla should forbid someone from directly registering their interceptor with a container; there are valid reasons to do so.  Maybe more control is needed over how it is built.  Some container frameworks can be configured to not resolve types that have not been explicitly registered, even if it technically could build the instance.  There are valid reasons for that too, so someone with that container configuration would not be able to use Csla at all if Csla restricted it.

> * my own personal reason for this future change would be to minimise allocations by having singleton interceptors - but not if it caused a bottleneck in the application because I had to use locks for thread safety. This is why I'm eager for this to be a seperate issue for future consideration as there is possibly more to it than meets the eye.

The way we're currently using an interceptor, that would flat out not work for us.  We'd be unable to use the DI feature of Csla at all in that scenario, and until the Interceptor idea was presented in that other discussion thread, that was going to be how we resolved the production issues; back out of the code that used the DI features of Csla completely.  So I hear what you're saying, but I don't think its reasonable to force a particular lifetime management on Csla users.

**Comment by ajohnstone-ks** [comment]
> I'm not adverse to the idea that CSLA 6 will require a lot of types currently provided through config to be registered via DI. This is a breaking change release, and I'd rather buy into DI now, than try to cobble together a nightmare combination of old-style provider models and new-style DI models. That will only lead to pain and confusion for years to come!

I would rather just one place to setup things for Csla as well.  My concern with using DI for some of these things like interceptor and authorizers though is that Csla gives up the lifetime management of those objects.  I could see that becoming problematic to support.  Then issues come up on the forum because someone needs to use HierarchicalLifetimeManager instead of ContainerControlledLifetimeManager (for example).  

And also using DI for those kinds of things is low on my list of wants for DI & Csla; I'd get much more value out of being able to inject dependencies into business & auth rules.

**Comment by dazinator** [comment]
> So I hear what you're saying, but I don't think its reasonable to force a particular lifetime management on Csla users.

Just to clarify, I wasn't suggested anything about `forcing a lifetime`. I'm just saying that if CSLA delegates lifetime to the container, there is nothing stopping users from registering (or overriding registrations for) their service with different lifetimes such as `singleton`. Think we are both on the same wavelength with why that might be potentially problematic, or beneficial - depending on what the service does and how its written.


If we did want to leverage DI, but try to limit the problem of misconfiguration, one idea is csla could offer registration extension methods that enforce the lifetime to what it expects, something like this:

```csharp
services.AddCsla().RegisterDataPortalInterceptor<MyCustomInterceptor>();
```

Where the extention method would do something like:

```
public static ICslaServicesBuilder  RegisterDataPortalInterceptor<TInterceptorType>(this ICslaServicesBuilder builder)
where TInterceptorType: IInterceptor
{
builder.Services.AddTransient<IInterceptor, TInterceptorType>();
}

public static ICslaServicesBuilder  RegisterDataPortalInterceptor<TInterceptorType>(this ICslaServicesBuilder builder, Func<IServiceProvider, TInterceptorType> factoryFunc)
where TInterceptorType: IInterceptor
{
builder.Services.AddTransient<IInterceptor, TInterceptorType>(factoryFunc);
}

```

That way, it would enable removal of those static properties where users configure `Type`'s, csla would then resolve like so:

```
var interceptor = DefaultServiceProvider.GetService<IInterceptor>();

```

In this case, as interceptors should be registered by the provided extension method this helps ensure they are registered with the supported lifetime - i.e transient.

If users really wanted to break the system by performing a manual registration as a singleton, they could do so, but that's true of pretty much anything in asp.net core.

**Comment by ajohnstone-ks** [comment]
> Just to clarify, I wasn't suggested anything about `forcing a lifetime`. I'm just saying that if CSLA delegates lifetime to the container, there is nothing stopping users from registering (or overriding registrations for) their service with different lifetimes such as `singleton`. Think we are both on the same wavelength with why that might be potentially problematic, or beneficial - depending on what the service does and how its written.

Yes, it sounds like we are in agreement.  It could be very powerful, and also very disastrous.  

> [extension method stuff]

I like this idea.  This could be the recommended way to configure an interceptor, but it also provides a way for someone to alter the lifetime or construction in a more advanced scenario too via the manual registration.

If there's something where its really important for Csla to retain control of the lifetime, maybe in that case it could reflect over the ctors of the given type (so a configuration property would still be required) and use `Activator.CreateInstanc(Type, params object[] args)`   Maybe that's a solution for certain types as well.

**Comment by dazinator** [comment]
> If there's something where its really important for Csla to retain control of the lifetime, maybe in that case it could reflect over the ctors of the given type (so a configuration property would still be required) and use `Activator.CreateInstance(Type, params object[])` Maybe that's a solution for certain types as well.

 - yeah this is done by a very simple switch out of the existing `Activator.CreateInstance` call for [ActivatorUtilities.CreateInstance](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.activatorutilities.createinstance?view=dotnet-plat-ext-5.0) which does the same but injects dependencies and was my original suggestion because it enables DI, doesn't break anything or require the interceptor type to be registered.

I think I have provided all the input I can on this now, and I guess we just need a decision on whether to make the simple change first to enable dependencies when activating the type (ActivatorUtilities.CreateInstance), or to go straight to requiring a DI registration and having csla resolve the type from DI rather than activate it.

**Comment by rockfordlhotka** [comment]
Right now I've done the simple change, though it isn't complete yet - but in _some_ cases the [instances are now created via DI (in a PR)](https://github.com/rockfordlhotka/csla/commit/7262a2ccae721a926d9c72a280045a04f794c3e9).

**Comment by ajj7060** [comment]
@rockfordlhotka I can take this up if the defaultinterceptor hasn't been created yet.

**Comment by rockfordlhotka** [comment]
Thank you @ajj7060, no, I haven't gotten to this yet.

The `main` branch now has the broader DI change in place, so most object instance creation is now done via DI (though in a different conversation it occurred to me that I need to identify all uses of the `new` keyword and consider updating each of them for DI too).

I haven't changed how `ApplicationContext` works yet, nor the config subsystem, so you should have a pretty straightforward time creating this new interceptor.

**Comment by rockfordlhotka** [comment]
@ajj7060 and @ajohnstone-ks and @dazinator 

Today I _finally_ merged the PR that gets the core `Csla` project building and (at least basically) running with dependency injection #2359. 

This included major rework to the way the data portal works, and it _might_ mean that _this_ issue to create a default interceptor is no longer required. I'm not sure, so I thought I'd provide my thoughts.

The new reliance on DI by CSLA means that the server-side data portal is now created as a service via DI. As such, it is created by the "current" provider being used by aspnetcore - just like a Page or Controller or any other service. All types subsequently used or created by the data portal are created based on that same DI provider.

As I understand it, aspnetcore creates a provider for, say, an http request, and then disposes that provider at the end of the call. That way they support scoped instances of services.

Because the server-side `DataPortal` object is created using this provider, it and everything it creates (if scoped) will be disposed when the call completes.

I think this was the intent of creating the default intercepter - and so I think this issue is no longer required.

Do you concur?

**Comment by dazinator** [comment]
@rockfordlhotka nice one - yes I concur. This was a stop gap solution in my view, to try to allow a DI scope to be "honoured" for server side DP operations via an extensible hook. If the dependency on DI is now in the core, and server side DP is itself a scoped service, and uses the current scope when activating business objects - all should be good like you say.

Side thought: If the server side DP is now itself a scoped service (analogous to an MVC controller or other endpoint) then it will be created and destroyed per request (or more accurately, per scope). Prior to these changes I am guessing it was essentialy a static or a singleton? This means the perf profile has changes slightly as there will be more work for the GC and there is potential for unnecessary repeated work. Depending on the level of optimisation you might want to investigate making the DP a singleton service rather than a scoped service so that it's one less object for the GC per request and keeps a similar perf profile to before these changes just with DI now in the loop which I think was the main desired feature

**Comment by rockfordlhotka** [comment]
The server side data portal has always been per-request to help ensure isolation.

**Comment by ajohnstone-ks** [comment]
> The new reliance on DI by CSLA means that the server-side data portal is now created as a service via DI. As such, it is created by the "current" provider being used by aspnetcore - just like a Page or Controller or any other service. All types subsequently used or created by the data portal are created based on that same DI provider.
> 
> As I understand it, aspnetcore creates a provider for, say, an http request, and then disposes that provider at the end of the call. That way they support scoped instances of services.

I'm not entirely sure if it does or not.  And I'm not sure if we're only talking about a remote application host, or any Asp.net application, where I'm writing controller code that calls BOs.

If we're talking about the former, my concern is that under high load scenarios, this leaves the connection open (and unavailable) after the BO has finished all its work, but while the Asp.net plumbing code is running (and probably not using it, and probably doesn't want to share the connection/transaction anyway with asp.net code using DBs if one is used).   At a former job, we did do a single connection per request and had to abandon it because of those issues.  We had enough requests to justify 12 web servers just for this app to give an idea of the normal load.

If we're talking about a full Asp.net app, I think the connection per request isn't workable at all; anytime you use more than one root BO in an action method, those BOs will end up sharing the connection and transactions get squirrely.  In this scenario every root BO call should get its own scope which is destroyed just before returning to the controller method.

In the application  host mode, the scope per web request will only create problems under load; in the logical 3-tier, things show up much more easily.

If the interceptor idea is rejected, I hope that the option to override the behavior with an interceptor is retained, because both of the above issues presented in our applications & application host and this is a serious enough issue for us that we'd have to consider abandoning Csla if we can't.
